src/core/cgroup.c

```diff
@@ -1888,8 +1888,12 @@ static int unit_update_cgroup(
          * We perform migration also with whole slices for cases when users don't care about leave
          * granularity. Since delegated_mask is subset of target mask, we won't trim slice subtree containing
          * delegated units.
+         *
+         * If we're in an nspawn container and using legacy cgroups, the controller hierarchies are mounted
+         * read-only into the container. We skip migration/trim in this scenario since it would fail
+         * regardless with noisy "Read-only filesystem" warnings.
          */
-        if (cg_all_unified() == 0) {
+        if (cg_all_unified() == 0 && !path_is_read_only_fs(u->cgroup_path)) {
```

Comment:
What about downgrading to log_unit_debug_errno if we get -EROFS? That way we still log something if we are in an odd situation. And it's a more common pattern as far as I can see.

Reply:
Updated to drop logging level to debug if the returned error is -EROFS.